[backport] SI-7470 implements fundep materialization

xeno-by opened the pull request:
Backports 21a8c6c from the 2.11.x branch under -Xfundep-materialization
as per Miles Sabin's request. Thanks Miles!

xeno-by committed:
[backport] SI-7470 implements fundep materialization

Backports 21a8c6c from the 2.11.x branch under -Xfundep-materialization
as per Miles Sabin's request. Thanks Miles!

xeno-by commented:
review @adriaanm @retronym @milessabin

xeno-by commented:
The original pull request for 2.11.x: https://github.com/scala/scala/pull/2821

adriaanm commented:
Thanks, thinking out loud: would it make sense to use -Xsource:2.11 for this instead. It doesn't really matter for 2.10, but in 2.11 we could centralize these things this way.

milessabin commented:
@xeno-by awesome ... thanks!

@adriaanm -Xsource:2.11 would work just as well for me, but what about -Xfuture?

adriaanm commented:
I'd like to subsume -Xfuture by -Xsource:$currentVersion.next (because we already do the latter for $currentVersion.prev)

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -886,10 +908,12 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
   private def isDelayed(expandee: Tree) = delayed contains expandee
   private def calculateUndetparams(expandee: Tree): scala.collection.mutable.Set[Int] =
-    delayed.get(expandee).getOrElse {
+    if (forced(expandee)) scala.collection.mutable.Set[Int]()
```
please add a comment: `!settings.XfundepMaterialization.value implies forced.isEmpty`

xeno-by replied:
done

adriaanm commented:
Whatever we decided, I'd rather not introduce another -X setting. I'd prefer -Xsource, but -Xfuture is fine too. (I forget whether 2.10 already has -Xsource.)

xeno-by committed:
pull request feedback

xeno-by commented:
I don't mind using any of the flags. Speaking of -Xsource:2.11, is there any precedent of it being used for something else?

adriaanm commented:
It's used in the 2.11 codebase to be backwards compatible with 2.10.

xeno-by commented:
Yeah, I saw that -Xsource:2.10 is used, but what about -Xsource:2.11? Is the particular 2.11 value used anywhere?

adriaanm commented:
I don't think so. We should systematically use it in 2.12 for any source
backwards incompatible changes -- conversely for 2.10

On Thu, Jul 3, 2014 at 1:24 PM, Eugene Burmako notifications@github.com
wrote:

> Yeah, I saw that -Xsource:2.10 is used, but what about -Xsource:2.11? Is
> the particular 2.11 value used anywhere?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/scala/scala/pull/3859#issuecomment-47894746.

xeno-by commented:
Allright, then we set the precedent and go for -Xsource:2.11 now?

adriaanm commented:
I think that would be more consistent. I won't insist too much, since 2.10.5 will be the last 2.10 release anyway.

xeno-by commented:
Funnily enough, we don't even have -Xsource in 2.10.x. How about we just use -Xfundep-materialization then? :)

adriaanm commented:
Sure. I suspected as much, but didn't have time to check.

gkossakowski commented:
@xeno-by, pinging on addressing @adriaanm's feedback. It seems like `-Xfundep-materialization` is the way to go, right?

xeno-by commented:
Looks like it, yes. I think I've addressed everything btw.

gkossakowski commented:
True. Thinking about it more, could we hide this behind 'Y' flag instead? We have lesser obligation to keep around `Y` flags and this is something we should remove from 2.11/2.12.

xeno-by commented:
We could do that, yes. Shall I rename the flag then? Any other comments that I could fix while I'm at it?

gkossakowski commented:
No, I think that's all I have. I don't understand the feature itself but I trust that you and @milessabin has spent enough of time on it :-)

Once you change the flag, ping me and I'll merge it.

xeno-by committed:
-Xfundep-materialization => -Yfundep-materialization

To quote gkossakowski:

  Thinking about it more, could we hide this behind 'Y' flag instead?
  We have lesser obligation to keep around Y flags and this is something
  we should remove from 2.11/2.12.

xeno-by commented:
Ping

xeno-by commented:
Ping @gkossakowski. Shall we merge?

gkossakowski commented:
Yep.

milessabin commented:
Thanks.
